modules/docs/markdown/06-guides/smithy-transformations.md

```diff
@@ -0,0 +1,124 @@
+
+@trait(selector: "structure > member")
+structure removeBeforeCodegen {}
+
+structure MyStruct {
+    @required
+    name: String
+
+    @removeBeforeCodegen
+    id: String
+}
+```
+
+Here we have defined a trait `removeBeforeCodegen`. We have marked the `id` member of `MyStruct` with this trait. As such, we will implement a transformer which will lead to the model looking as follows:
+
+```smithy
+namespace preprocessed
+
+structure MyStruct {
+    @required
+    name: String
+}
+```
+
+This is the model that will ultimately be fed into the Smithy4s code generation tooling.
+
+## Implement the Transformer
+
+Now we will walk through the steps of actually implementing the Smithy model transformer. The `smithy-build` Java library provides an interface for these called `ProjectionTransformer`. Here is an example of a transformer that will remove the members marked with the `removeBeforeCodegen` trait as discussed above.
+
+```scala
+final class RemoveBeforeCodegenTransformation extends ProjectionTransformer {
+  def getName() = {
+    "RemoveBeforeCodegenTransformation"
+  }
+
+  def transform(x: TransformContext) = {
+    val toRemove = x
+      .getModel()
+      .getShapesWithTrait(new ToShapeId() {
+        override def toShapeId(): ShapeId =
+          ShapeId.from("preprocessed#removeBeforeCodegen")
+      })
+    x.getTransformer().removeShapes(x.getModel(), toRemove)
+  }
+}
+```
+
+Here we create a `ProjectionTransformer` named `RemoveBeforeCodegenTransformation`. Inside the `transform` method we remove all shapes that are marked with the `removeBeforeCodegen` trait and return the resultant model.
+
+## Register the Transformer
+
+We need to register the Transformer so that the Smithy tooling will be able to find it when it does a lookup for all `ProjectionTransformer` implementations. We do this by creating a file at `src/main/resources/META-INF/services/` called `software.amazon.smithy.build.ProjectionTransformer`. Inside of this file, we need to place the fully qualified classpath that leads to our transformer (e.g. `preprocessed.transformations.RemoveBeforeCodegenTransformation`).
+
+## Wire up in SBT
+
+In our `build.sbt` file we will create a new project called `transformers` that looks as follows:
+
+```scala
+lazy val transformers = (project in file("transformers"))
+  .settings(
+    name := "transformers",
+    libraryDependencies += "software.amazon.smithy" % "smithy-build" % smithy4s.codegen.BuildInfo.smithyVersion
+  )
+```
+
+This project is where our `RemoveBeforeCodegenTransformation` will live. As such, it requires the `smithy-build` dependency so it can access the `ProjectionTransformer` interface.
+
+Now, in our project that is using the Smithy4s SBT plugin (`.enablePlugins(Smithy4sCodegenPlugin)`) we need to add the following settings:
+
+```scala
+...
+.enablePlugins(Smithy4sCodegenPlugin)
+.settings(
+  ...
+  // transformers
+  Compile / smithy4sModelTransformers += "RemoveBeforeCodegenTransformation",
```

Comment:
Worth mentioning again that this is the name returned by the method

Reply:
maybe we can also change this name so that it's clearly different from the class name? sth like `remove-before-codegen`. I'm not sure if there's any convention for this.